onadata/libs/serializers/project_invitation_serializer.py

```diff
@@ -0,0 +1,129 @@
+import re
+from django.conf import settings
+from django.utils.translation import gettext as _
+from django.contrib.auth import get_user_model
+from django.utils import timezone
+from rest_framework import serializers
+from onadata.apps.logger.models import ProjectInvitation
+from onadata.libs.permissions import ROLES
+
+
+User = get_user_model()
+
+
+class ProjectInvitationSerializer(serializers.ModelSerializer):
+    """Serializer for ProjectInvitation model object"""
+
+    def validate_email(self, email):
+        """Validate `email` field"""
+        # Regular expression pattern for email validation
+        pattern = r"^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$"
+        err_msg = "Invalid email."
+
+        # Check if the email matches the pattern
+        if not re.match(pattern, email):
+            raise serializers.ValidationError(_(err_msg))
+
+        domain_whitelist = getattr(
+            settings, "PROJECT_INVITATION_EMAIL_DOMAIN_WHITELIST", []
+        )
+
+        if domain_whitelist:
+            # Extract the domain from the email address
+            domain = email.split("@")[1]
+
+            # Check if the domain matches "foo.com"
+            if not domain.lower() in [
+                allowed_domain.lower() for allowed_domain in domain_whitelist
+            ]:
+                raise serializers.ValidationError(_(err_msg))
+
+        # email should not be of an existing user
+        if User.objects.filter(email=email).exists():
+            raise serializers.ValidationError(_("User already exists"))
+
+        return email
+
+    def validate_role(self, role):
+        """Validate `role` field"""
+        if role not in ROLES:
+            raise serializers.ValidationError(_("Invalid role."))
+
+        return role
+
+    def get_unique_together_validators(self):
+        """Overriding method to disable unique together checks"""
+        return []
```

Comment:
Why ?

Reply:
I am using the  POST endpoint for creating and updating an invitation. I had challenges validating data for both create and update. This was a workaround to have the serializer validation work correctly when updating the invitation. 

It wasn't a good workaround and I have been thinking of the best way to solve it. I have updated the code with a better implementation by introducing `ProjectInvitationBaseSerializer`

Reply:
I'm against using the `POST` endpoint for updates... Updates should be correctly tied to either a `PATCH` or `PUT` request

Reply:
@DavisRayM Ok...I had separated them but thought it would be an overkill for this particular implementation. But no problem we can use PATCH / PUT